MVP v1.0.0

emcniece opened the pull request:
This branch is now ready for production testing.

SSL cert generation and Nginx templating work properly. There is no file distribution yet, so multi-host environments are not supported.

emcniece committed:
Adding templates and docker-compose

emcniece committed:
Fixing duplicate add

emcniece committed:
Fixing Dockerfile, docker-compose

emcniece committed:
Adding Makefile, VERSION

emcniece committed:
Allowing default rancher-gen config to be overridden by CMD

emcniece committed:
Removing CMD because it doesn't work

emcniece committed:
Fixing volumed rancher-gen.cfg location

Commit:
used rancher-gen.cfg during container start

emcniece committed:
Updating template & readme with better tags names

emcniece committed:
Merge branch 'dev' of https://github.com/CausticLab/rgon-proxy into dev

Commit:
remove falsly added config

Commit:
Merge branch 'dev' of github.com:CausticLab/rgon-proxy into dev

emcniece committed:
v0.0.3: Fixing acmetook unpack

Commit:
fixed paths and added proper docker-compose logging

Commit:
Merge branch 'dev' of github.com:CausticLab/rgon-proxy into dev

Fixed paths
Added proper logging in docker-compose.yml

Commit:
implement new port rule as discussed in discord

emcniece committed:
Adding acmetool proxy setup

emcniece committed:
Merge branch 'dev' of https://github.com/CausticLab/rgon-proxy into feat/acmetool-proxy

Commit:
use new v0.3.0 version of rancher-gen

emcniece committed:
Merge pull request #1 from CausticLab/feat/acmetool-proxy

Adding acmetool proxy setup

emcniece committed:
Fixing Nginx template port lists

emcniece committed:
Incrementing rancher-gen version

emcniece committed:
Adding acmetool hooks for cert/rancher-gen handling

emcniece committed:
Updating go-rancher-gen to 0.4.3

emcniece committed:
First entrypoint.sh version

emcniece committed:
Revising first-run process

emcniece committed:
Rancher-Gen 0.5.0, startup testing, minor cleanup

emcniece committed:
Nginx tpl supportin HTTPS redirect

Munsio commented on app/entrypoint.sh:
```diff
@@ -0,0 +1,63 @@
+function check_dh_group {
+  if [[ ! -f /etc/nginx/dhparam/dhparam.pem ]]; then
+    echo "Creating Diffie-Hellman group (can take several minutes...)"
+    openssl dhparam -out /etc/nginx/dhparam/.dhparam.pem.tmp 2048
```
should we make this param configurable?

emcniece replied:
We certainly could. Currently [entrypoint.sh](https://github.com/CausticLab/rgon-proxy/pull/2/files#diff-f4c0396de4314fb0fd536fae34d42438) generates this `dhparam.pem` file in the `check_dh_group` function, so we would need to use parameters in both of these places.

How do you propose parameterization - environment variable?

Munsio replied:
i would suggest environment vars

emcniece replied:
https://github.com/CausticLab/rgon-proxy/issues/3

Munsio commented on examples/acmetool/hooks/01triggerRancherGen.sh:
```diff
@@ -0,0 +1,16 @@
+    mkdir -p /etc/nginx/certs
+    cp -RLf /var/lib/acme/live/* /etc/nginx/certs/
+    rancher-gen --config ${RGEN_CONFIG_ONETIME}
+    ;;
```
should we copy the files or write them directly into the /nginx/certs folder?
we also could symlink the folders

emcniece replied:
First, this file is actually deprecated and should now be removed! This hook was introduced to get Rancher-Gen to re-run after certificates were loaded, but it created an extra loop iteration that was not desired.

Old loop:

1. Stack created
1. Rancher-Gen.cfg tpl 1 - generate basic nginx.config (allows nginx containers to come online properly).
1. Rancher-Gen.cfg tpl 2 - Create SSL certs via `notify-cmd="acmetool want {{rgon.domain}}`
1. 01triggerRancherGen.sh - copy certs to destination
1. Rancher-Gen-onetime.cfg - generate new nginx.config with SSL certs
1. Repeat from step 2

So each time the main Rancher-Gen.cfg runs (which is frequently, not _just_ when metadata updates) it would trigger acmetool and reload the `nginx` containers twice.

A better approach is to generate the SSL certs first, then reload `nginx` containers once. However this creates a race condition in that the `nginx` containers can't start immediately as they have to wait for `acmetool` to do an initial cert run. The solution is to run a basic Rancher-Gen pass to create `nginx.conf` quickly, then allow the Rancher-Gen SSL run.

New loop:

1. Stack created
1. Rancher-Gen-firstrun.cfg - generate basic nginx.config (allows nginx containers to come online properly). Only runs once on stack creation.
1. Rancher-Gen.cfg tpl 1 - Create SSL certs
1. Rancher-Gen.cfg tpl 2 - generate new nginx.config with SSL certs
1. Repeat from step 3

Finally, Nginx can't read certificates from symlinked directories - the certs need to be `cp -L` (dereferenced) before use.

emcniece replied:
This "new loop" is implemented already - 01triggerRancherGen.sh is deprecated and can be removed.

emcniece replied:
https://github.com/CausticLab/rgon-proxy/issues/5

Munsio commented on examples/acmetool/responses:
```diff
@@ -0,0 +1,15 @@
+"acme-agreement:https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf": true
+"acmetool-quickstart-choose-server": https://acme-staging.api.letsencrypt.org/directory
+"acmetool-quickstart-choose-method": proxy
+"acmetool-quickstart-install-cronjob": true
```
does this even work on an alpine image? i thought they have no crontab

emcniece replied:
Great observation. This is untested - these are default options from the sample `responses` file.

Munsio replied:
i think we need to install cron and then run it in the entrypoint.sh file

emcniece replied:
We will also need to use entrypoint.sh ENV replacement here on `acme-staging.api` to `acme-v01.api`.

emcniece replied:
- https://github.com/CausticLab/rgon-proxy/issues/3
- https://github.com/CausticLab/rgon-proxy/issues/4

Munsio commented on examples/acmetool/responses:
```diff
@@ -0,0 +1,15 @@
+# You will typically want to use --response-file with --stdio or --batch.
+# For dialogs not requiring a response, but merely acknowledgement, specify true.
+# This file is YAML. Note that JSON is a subset of YAML.
+"acme-enter-email": "hostmaster@example.com"
```
should be an customizeable email-adress

emcniece replied:
Haven't found a solution for this yet. https://github.com/hlandau/acme/issues/243

From [acme/register.go:74](https://github.com/hlandau/acme/blob/83d2466ceb01a76d281ff8983465db191fbfa1e2/solver/register.go#L74): "... Although entering an e. mail address is optional, it is highly recommended."

Munsio replied:
maybe we need to generate this file in the entrypoint.sh on each startup? With the rancher-catalog we can pass possible vars as env-vars so this should be no problem.

emcniece replied:
Agreed, `entrypoint.sh` for ENV regex substitution.

emcniece replied:
https://github.com/CausticLab/rgon-proxy/issues/3

Munsio commented on examples/rancher-gen/nginx.tmpl:
```diff
@@ -0,0 +1,183 @@
+
+# If we receive X-Forwarded-Proto, pass it through; otherwise, pass along the
+# scheme used to connect to this server
+map $http_x_forwarded_proto $proxy_x_forwarded_proto {
```
could you explain why we need this?

emcniece replied:
1. Don't need Host block (lines 1-7), this was for debug/ID
1. Access log formatting needed for tracking traffic source/dest through the Nginx container
1. Not sure about X-Forwarded-Proto, taken from [docker-gen](https://github.com/jwilder/docker-gen)

emcniece replied:
https://github.com/CausticLab/rgon-proxy/issues/6

Munsio commented on examples/rancher-gen/nginx.tmpl:
```diff
@@ -0,0 +1,183 @@
+  '' close;
+}
+
+gzip_types text/plain text/css application/javascript application/json application/x-javascript text/xml application/xml application/xml+rss text/javascript;
```
gzip compression should be possible to turn on/off and set the specific file-types cause sometimes this can break things

emcniece replied:
Two proposals:

1. Move `gzip_types` to `server{}` directives ([docs](http://nginx.org/en/docs/http/ngx_http_gzip_module.html)) ([example](https://github.com/jwilder/docker-gen/blob/master/templates/nginx.tmpl#L46))
1. Implement [vhost overriding](https://github.com/Munsio/docker-letsencrypt-nginx-proxy-companion/blob/master/app/nginx.tmpl#L186-L190)

emcniece replied:
https://github.com/CausticLab/rgon-proxy/issues/7

Munsio commented on examples/rancher-gen/nginx.tmpl:
```diff
@@ -0,0 +1,183 @@
+{{end}}
+
+## HOSTS ----------------
+{{range $labelValue, $hosts := hosts | groupByLabel "rgon.domain" -}}
```
i dont know if it is a great idea to match a domain to an host.
Is there an use-case for it? Using rgon-proxy without having the app in an container maybe? But then why should we support this if we say that we are a special proxy service for rancher.

emcniece replied:
I use this to proxy Rancher itself! Adding labels to the main Rancher host allows us to access Rancher at https://rancher-test.com instead of http://rancher-test.com:8080.

emcniece replied:
https://github.com/CausticLab/rgon-proxy/issues/8

Munsio reviewed (changes requested):
review *yeah*

emcniece committed:
Removing hook trigger file (#10)

emcniece committed:
Removing nginx html directory (#11)

Munsio committed:
Fix #12 by changing dhparam folder to certs folder

Munsio committed:
use dir/* to copy multiple files

Munsio committed:
Update Dockerfile

try another approach with COPY instead of ADD

Munsio committed:
Update Dockerfile

copy example files to /app directory and copy them again in the entrypoint.sh to the correct directory path cause volume generations overrides the added files

Munsio committed:
Update entrypoint.sh

copy config files from app to default directory if the directory not exists

Munsio committed:
Update entrypoint.sh

fixed if condition

emcniece committed:
Adding cert copy hook, fixing docker-compose.yml.sample

emcniece committed:
Removing nginx.tmpl debug header

emcniece committed:
Manual cron install for Acmetool (#14)

emcniece committed:
Add ability to swap Acmetool vars with ENV (#15)

emcniece committed:
Fix/Nginx Cert Templates (#22)

* Fixing hook perms, updating nginx.tmpl vars

* Nginx.tmpl update, adding default certs

Commit:
Add first templating features

emcniece committed:
Incrementing rancher-gen to 0.6.0

Commit:
modified nginx and rancher-gen template to support multiple domains for ssl-generation

Commit:
removed not wanted changes

Commit:
fixed brackets

Commit:
Merge remote-tracking branch 'origin/dev' into feature/template-templating

Commit:
Move the server blocks and all that things into templates

Commit:
Merge remote-tracking branch 'origin/feature/template-templating' into feature/multi-domain-ssl

Commit:
Modified check-cmd to use first domain from label to check for existing
certificates

emcniece committed:
Fixing / 301 redirect

Munsio committed:
Merge pull request #23 from CausticLab/feature/multi-domain-ssl

Feature/multi domain ssl

Munsio committed:
Merge pull request #27 from CausticLab/feature/template-templating

Feature/template templating

Munsio committed:
Update README.md

Munsio committed:
Update README.md

Munsio committed:
Update README.md

Commit:
fixed #33

Commit:
Add current feature list and upcoming features

Commit:
typo fixes

Munsio committed:
Added the possibility to enable the stats page

Munsio committed:
Merge pull request #34 from CausticLab/feature/hardcoded-acmetool-port

fixed #33

emcniece committed:
Adding usage instructions, including rgon.stats

Munsio committed:
Update README.md

added url where the stats can be viewed

Munsio committed:
fixes #38
added acme-challenge location in https server block

Munsio committed:
Merge pull request #39 from CausticLab/bugfix/https-acme-challenge

fixes #38

Munsio committed:
Merge pull request #37 from CausticLab/feature/nginx-stats

Added the possibility to enable the stats page

Munsio committed:
Merge remote-tracking branch 'origin/dev' into bugfix/https-acme-challenge

Munsio committed:
fixes #45 hopefully

Munsio committed:
Merge pull request #46 from CausticLab/bugfix/cronjob-reconcile

Bugfix/cronjob reconcile

Munsio committed:
added missing volume attribute so content don't get overwritten

Munsio committed:
adding the volume was wrong fix it by copying the config files into directory if they dont exist

Munsio committed:
fixes #47

added an mkdir -p to the copy stmnt of the response - shouldn't do anything if the directory already exists and fix a problem with cp if folder doesnt exist

Munsio committed:
Update responses

changed the url to the new acme TOS pdf

Munsio committed:
Add location block for 301 redirect under http
Update Version to 0.1.0

Munsio committed:
update version again

Munsio committed:
Merge branch 'master' into dev
